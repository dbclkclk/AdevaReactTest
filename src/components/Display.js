import React from "react";
import { Button, Form, FormGroup, Label, Input, Alert} from 'reactstrap';
import NoteAction from "../actions/note";
import NoteStore from "../stores/notestore";

export default class Display extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
        form: {
            note: "",
            id: null
        }, 
        saved: false
    }
    this.submitForm = this.submitForm.bind(this);
    this.handleState = this.handleState.bind(this);
    this.loadNote = this.loadNote.bind(this);
  }

  componentWillMount() {
      NoteStore.addChangeListener(this.loadNote);
  }
  loadNote() {
      let note = NoteStore.getState();
      this.setState({form: {note:note.value, id: note.key}});
  }
  submitForm(e) {
      e.preventDefault();
      if (this.state.form.id==null) {
        NoteAction.Add(this.state.form.note);
      }
      else 
      {
        NoteAction.Update(this.state.form.id, this.state.form.note);
      }
      let form = {...this.state.form};
      form.note = "";
      form.id = null;
      this.setState({form:form, saved: true});
  }
  handleState(e) {
    let form = {...this.state.form};
    form[e.target.name] = e.target.value;
    this.setState({form: form, saved: false});
  }
  render() {
      let message = this.state.saved === true ? (<Alert color="success">
                Note Saved!
            </Alert>) : "";
      return (
        <div className="Container-fluid">
            <h3>Add/Edit Notes</h3>
            {message}
            <Form onSubmit={this.submitForm}>
                <FormGroup>
                    <Input type="textarea" name="note" id="note" onChange={this.handleState} placeholder="Enter your note here" value={this.state.form.note} />
                </FormGroup>
                <Button>Submit</Button>
            </Form>
        </div>
      );
  }
}
