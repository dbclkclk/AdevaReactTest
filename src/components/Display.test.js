import React from "react";
import { shallow, mount } from 'enzyme';
import Display from "./Display";
import NoteStore from "../stores/notestore";
import NoteAction from "../actions/note";

let mockaddChangeListener = jest.fn();
NoteStore.addChangeListener = (fn) => mockaddChangeListener = fn;
let mockgetState = jest.fn();
NoteStore.getState = mockgetState;

let mockAdd = jest.fn();
let mockUpdate = jest.fn();
NoteAction.Add = mockAdd;
NoteAction.Update = mockUpdate;

describe("Display Component", () => {
  it("it should render initial dom", () => {
    const wrapper = shallow(<Display />);
    expect(wrapper.getElements()).toMatchSnapshot();
  });
  it("Test that when note change, loads data from store and display it in DOM", () => {
    const component = mount(<Display />);
    mockgetState.mockReturnValue({value:"blah blah", key: null});
    mockaddChangeListener();
    component.update();
    expect(component.find("#note").at(1).props().value).toBe("blah blah");
  });
  it("Test that when note is being added, it adds it", () => {
    const component = mount(<Display />);
    component.find("#note").at(0).simulate('change', {target: {value: 'my value', name: 'note'}});
    component.find("#note").at(1).simulate('change', {target: {value: 'my value', name: 'note'}});
    const form = component.find('form').at(0);
    form.simulate('submit');
    expect(mockAdd.mock.calls[0][0]).toBe("my value");
  });
  it("Test when note is being edited, it calls update", () => {
    const component = mount(<Display />);
    mockgetState.mockReturnValue({value:"blah blah", key: 1});
    mockaddChangeListener();
    component.update();
    component.find("#note").at(0).simulate('change', {target: {value: 'new value', name: 'note'}});
    component.find("#note").at(1).simulate('change', {target: {value: 'new value', name: 'note'}});
    const form = component.find('form').at(0);
    form.simulate('submit');
    expect(mockUpdate.mock.calls[0][0]).toBe(1);
    expect(mockUpdate.mock.calls[0][1]).toBe("new value");
  });
  
});