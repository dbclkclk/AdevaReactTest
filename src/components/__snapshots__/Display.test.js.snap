// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Component Test that when componentWillMount, loads data from store and display it in DOM 1`] = `
Array [
  <Display />,
]
`;

exports[`Display Component it should render initial dom 1`] = `
Array [
  <div
    className="Container-fluid"
  >
    <h3>
      Add/Edit Notes
    </h3>
    
    <Form
      onSubmit={[Function]}
      tag="form"
    >
      <FormGroup
        tag="div"
      >
        <Input
          id="note"
          name="note"
          onChange={[Function]}
          placeholder="Enter your note here"
          type="textarea"
          value=""
        />
      </FormGroup>
      <Button
        color="secondary"
        tag="button"
      >
        Submit
      </Button>
    </Form>
  </div>,
]
`;

exports[`Display Component it should render initial dom 2`] = `
Array [
  <div
    className="Container-fluid"
  >
    <h3>
      Add/Edit Notes
    </h3>
    
    <Form
      onSubmit={[Function]}
      tag="form"
    >
      <FormGroup
        tag="div"
      >
        <Input
          id="note"
          name="note"
          onChange={[Function]}
          placeholder="Enter your note here"
          type="textarea"
          value=""
        />
      </FormGroup>
      <Button
        color="secondary"
        tag="button"
      >
        Submit
      </Button>
    </Form>
  </div>,
]
`;

exports[`Display Component it should render initial state 1`] = `
Array [
  <div
    className="Container-fluid"
  >
    <h3>
      Add/Edit Notes
    </h3>
    
    <Form
      onSubmit={[Function]}
      tag="form"
    >
      <FormGroup
        tag="div"
      >
        <Input
          id="note"
          name="note"
          onChange={[Function]}
          placeholder="Enter your note here"
          type="textarea"
          value=""
        />
      </FormGroup>
      <Button
        color="secondary"
        tag="button"
      >
        Submit
      </Button>
    </Form>
  </div>,
]
`;
